### PR TITLE
Optimization of labels handling in issue_search

### PR DESCRIPTION
**Pull request by xlii-chl**
This PR enhances the labels handling in issue_search by optimizing the SQL query and de-duplicating the ids when generating the query string.

Those patches are backportable to v1.20.2 as is.

---

### Background

For some time now, BingBot and some other crawlers have been putting my Gitea instance on its knees with requests containing a lot of label ids, like this one :
```
[07/Aug/2023:11:28:37 +0200] "GET /Dolibarr/sendrecurringinvoicebymail/issues?q=&type=all&sort=&state=closed&labels=1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c2%2c10%2c2%2c1%2c1%2c10%2c10%2c7%2c6%2c10%2c10%2c3%2c2%2c1%2c5%2c10%2c1%2c6%2c2%2c7%2c3%2c7%2c6%2c10%2c1%2c10%2c1%2c1%2c7%2c7%2c1%2c1%2c1%2c1%2c10%2c10%2c1%2c2%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c1%2c2%2c1%2c12%2c6%2c6%2c10&milestone=0&project=-1&poster=0 HTTP/1.1" 499 0 "-" "Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko; compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm) Chrome/103.0.5060.134 Safari/537.36"
```

Since each of the label ids implies a join, it grows exponentially expensive for the database engine (at least on PostgreSQL but SQLite suffers a little too).

Thus this PR proposes two enhancements:
* rewrite the database query to use only one squashed condition,
* deduplicate the label ids when generating the URL.

This is my first try at Golang so don't hesitate to point mistakes or inelegant code.

As a side note, I unfortunately don't have the start of this snowball in my webserver logs anymore, and couldn't really reproduce it (the labels in the [above project](https://code.bugness.org/Dolibarr/sendrecurringinvoicebymail) don't have scopes, and the label id n°1 that we can see in the query string is linked to a deleted repository (but its labels still exist in database apparently)).

### CODE REVIEW

**Comment by xlii-chl** [comment]
Rewrote using new Slices.Sort & Compact methods from Go 1.21.

File: models/issues/label.go [hunk 1]
```diff
@@ -149,15 +149,26 @@ func (l *Label) LoadSelectedLabelsAfterClick(currentSelectedLabels []int64, curr
...
-		labelQuerySlice = append(labelQuerySlice, labelID)
+		labelQuerySlice = append(labelQuerySlice, l.ID)
 	}
 	l.IsSelected = labelSelected
-	l.QueryString = strings.Join(labelQuerySlice, ",")
+
+	// Sort and deduplicate the ids to avoid the crawlers asking for the
+	// same thing with simply a different order of parameters
+	slices.Sort(labelQuerySlice)
+	labelQuerySlice = slices.Compact(labelQuerySlice)
```
Comment (KN4CK3R) [inline models/issues/label.go:171]: Just use `code.gitea.io/gitea/modules/container.Set`. Then you don't need to sort and compact.

Comment (xlii-chl) [reply models/issues/label.go:171]: Thanks for the review ! However, I used container.Set in [my previous iteration ](https://github.com/xlii-chl/gitea/blob/5314cd1202a2c47a53bb0174b2417f658f721e42/models/issues/label.go#L151)and the code was less readable and the output's ordering was random (container.Set uses map).
I still use it in [models/issues/issue_search.go](https://github.com/go-gitea/gitea/pull/26460/files#diff-59d96577a83fb843b6b65a0777998803eba3212a88e37211d1b13b24f285134f) where its `Add()` method and the on-the-fly de-duplication are quite handy.

Comment (KN4CK3R) [reply models/issues/label.go:171]: Didn't see that the output is not just used in the backend. Then you could just use the set with `slices.Sort` at the end but that is only better if `labelQuerySlice` is very large.

**Comment by lunny** [comment]
Sorry missed this one. It seems interesting. I will take a look at next week.

**Review by lunny — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> Sorry, this PR got lost in the backlog. I'm currently going through the list of PRs still open and looking at getting them in.

But it seems that there are still some nits unresolved, including @KN4CK3R 's concern.

**Comment by wxiaoguang** [comment]
->  Refactor issue label selection #31497

File: models/issues/label.go (hunk 1 above)
Comment (wxiaoguang) [reply models/issues/label.go:171]: I think KN4CK3R 's suggestion is right. Managed to do some rewriting: #31497

**Comment by wxiaoguang** [comment]
@xlii-chl I saw that you said "weird ? The PR was sleeping for a year".

Since the hard-fork community is quite hostility to me, I could only answer your question here.

### Why this PR becomes stale?

Because the review suggestion is not resolved and you didn't respond for long time, and the code style doesn't quite match Golang standard. Indeed it needs more work.

If you could answer the review suggestion (either apply it or explain why not doing that), it could move on easily.

### What happened to the hard-fork?

Well, that's a long story, in short:

* They keeps criticizing Gitea community/maintainers and made a lot of lies, and they copied a lot of Gitea's feature and publicize that "they are their work" (for example: Gitea Actions).
* And be careful about their cherry-picks and PRs, I am sure there are not only one security problems (include XSS, permission bypass) in their codebase but not in Gitea.

----

* They keeps making false statements, for example: https://github.com/go-gitea/gitea/pull/27455#issuecomment-1753426680 , for the contributing guideline, it has been there for 7 years, they have been kindly warned not only one time, but you keeps saying that "Oh I don't know" and try to catch every chance to criticize Gitea.

* They keeps claiming that "their test is good" but "Gitea merge large untested code", but their PRs to Gitea didn't have tests, eg:  https://github.com/go-gitea/gitea/pull/27753 and https://github.com/go-gitea/gitea/pull/26841, and there are even more. There was no test, and even worse, they are incorrect. 

And here are why many of theirs PRs to Gitea were not merged mainly because there are bugs and they do not respond.

Let's see why some of them are not merged (I only picked some my reviews):

Really suggest everyone to take a deep look ........

* https://github.com/go-gitea/gitea/pull/28541#issuecomment-1926123191
* https://github.com/go-gitea/gitea/pull/28262#issuecomment-1879946412
* https://github.com/go-gitea/gitea/pull/28261#issuecomment-1831117073
* https://github.com/go-gitea/gitea/pull/28139#pullrequestreview-1740883374
* https://github.com/go-gitea/gitea/pull/27895#issuecomment-1793321693
* https://github.com/go-gitea/gitea/pull/27753#pullrequestreview-1692793142
* https://github.com/go-gitea/gitea/pull/27455#issuecomment-1753426680
* https://github.com/go-gitea/gitea/pull/27131#discussion_r1330327223
* https://github.com/go-gitea/gitea/pull/26841#issuecomment-1713876896
* https://github.com/go-gitea/gitea/pull/26488#issuecomment-1677094334
* https://github.com/go-gitea/gitea/pull/26487#issuecomment-1677110682

**Comment by lunny** [comment]
Let's not backport this since it's not a bug fix.

**Pull request merged**